### PR TITLE
fix rails 5 tests, fix more bugs, document more bugs

### PR DESCRIPTION
**Pull request by grosser**
- log_unpermitted just handles invalid ... renamed
 - arrays are not filtered when hash is requested
 - array elements are not filtered ... confirmed this bug exists on master

### CODE REVIEW

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb [hunk 1]
```diff
@@ -1,312 +1,274 @@
...
 require 'stronger_parameters/controller_support/permitted_parameters'
 
-SingleCov.covered! unless Rails::VERSION::MAJOR >= 5 && Rails::VERSION::MINOR > 0
+SingleCov.covered!
 
-describe StrongerParameters::ControllerSupport::PermittedParameters do
-  class WhitelistControllerTester < ActionController::Base
-    include StrongerParameters::ControllerSupport::PermittedParameters
+class WhitelistsController < ActionController::Base
+  ROUTES = ActionDispatch::Routing::RouteSet.new
```
Comment (dadah89) [inline test/stronger_parameters/controller_support/permitted_parameters_test.rb]: 👍

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb [hunk 2]
```diff
@@ -1,312 +1,274 @@
...
+    if Rails::VERSION::MAJOR < 5
+      super(action, options.fetch(:params).merge(format: options[:format] || 'html'))
     else
-      @controller.request.send(:deep_munge, params)
+      super
     end
   end
 
   before do
-    if Rails::VERSION::MAJOR >= 5 && Rails::VERSION::MINOR > 0
```
Comment (dadah89) [inline test/stronger_parameters/controller_support/permitted_parameters_test.rb]: 👍 👍

File: lib/stronger_parameters/controller_support/permitted_parameters.rb [hunk 3]
```diff
@@ -85,15 +85,15 @@ def permit_parameters
...
         log_unpermitted = self.class.log_unpermitted_parameters
+        permitted_params = without_invalid_parameter_exceptions(log_unpermitted) { params.permit(permitted) }
+        unpermitted_keys = flat_keys(params) - flat_keys(permitted_params)
 
         show_unpermitted_keys(unpermitted_keys, log_unpermitted)
 
         return if log_unpermitted
 
-        params.replace(permitted_params)
+        (ActionPack::VERSION::MAJOR >= 5 ? params.send(:parameters) : params).replace(permitted_params)
```
Comment (dadah89) [inline lib/stronger_parameters/controller_support/permitted_parameters.rb]: did you mean `params.send(:parameters, permitted_params)` ? if yes, is it feasible to get a failing test?

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb [hunk 4]
```diff
@@ -1,312 +1,274 @@
...
-      skip('PermittedParameters not compatible with Rails 5.1 or later')
-    end
-    WhitelistControllerTester.instance_variable_set(:@permit_parameters, nil)
-    @oldu = Parameters.action_on_unpermitted_parameters
-    Parameters.action_on_unpermitted_parameters = :log # same as dev/production
+    @routes = WhitelistsController::ROUTES
```
Comment (dadah89) [inline test/stronger_parameters/controller_support/permitted_parameters_test.rb]: not used?

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb [hunk 5]
```diff
@@ -1,312 +1,274 @@
...
 
-    describe 'with empty array' do
-      subject { { something: [] } }
-
-      it 'does not remove something' do
-        assert @controller.params.key?(:something)
-      end
+    # should behave exactly like submitting and empty array does ... but that's a bug in .permit
+    it 'does not removes emptied hash' do
+      do_request something: {remove: 1}
```
Comment (dadah89) [inline test/stronger_parameters/controller_support/permitted_parameters_test.rb]: does it make any difference to test `something: { test: { remove: 1 } }` instead? (guess it'd just nested, that's already being tests somewhere else)

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb [hunk 6]
```diff
@@ -1,312 +1,274 @@
...
+        "action" => "index",
+        "format" => "png",
+        "authenticity_token" => "auth"
+      )
+    end
 
-        assert !@controller.params.key?(:invalid)
-        refute @controller.response.headers['X-StrongerParameters-API-Warn'].nil?
-      end
+    it 'filters request.params' do
```
Comment (dadah89) [inline test/stronger_parameters/controller_support/permitted_parameters_test.rb]: this test and the previous one are exactly the same?

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb [hunk 7]
```diff
@@ -1,312 +1,274 @@
...
+      )
+    end
 
-      it 'filters nil values and send warn' do
-        @controller.params[:invalid] = nil
-        Rails.configuration.stronger_parameters_violation_header = 'X-StrongerParameters-API-Warn'
-        permit_without_raising
+    it "allows anything through, but warns" do
+      WhitelistsController.instance_variable_set(:@permit_parameters, nil)
+      WhitelistsController.permitted_parameters :index, :anything
```
Comment (dadah89) [inline test/stronger_parameters/controller_support/permitted_parameters_test.rb]: do we test anywhere that `:anything` doesn't strip parameters?

File: test/stronger_parameters/controller_support/permitted_parameters_test.rb [hunk 8]
```diff
@@ -1,312 +1,274 @@
...
+      end
 
-      refute @controller.params.key?(:everything)
-      assert @controller.params.key?(:something)
+      before { Parameters.action_on_invalid_parameters = :raise }
 
-      @controller.params.merge!(action: 'create', everything: 'bleh', something: 'hello')
-      permit_without_raising
+      it "raises" do
+        Parameters.action_on_invalid_parameters = :raise
```
Comment (dadah89) [inline test/stronger_parameters/controller_support/permitted_parameters_test.rb]: already done in the before block

**Comment by grosser** [comment]
one is for params the other is for request.params since we replace both

**Comment by grosser** [comment]
either one would not work since .permit does not clear empty

**Comment by grosser** [comment]
fails without

**Comment by grosser** [comment]
no, getting the internal hash from params since it's a
ActionController::Parameters code is executed when this runs on rails 5

**Comment by grosser** [comment]
others fixed

File: lib/stronger_parameters/controller_support/permitted_parameters.rb (hunk 3 above)
Comment (dadah89) [inline lib/stronger_parameters/controller_support/permitted_parameters.rb]: oh I see misread the code

**Pull request merged**